### PR TITLE
Remove DataProviders from Registrar and into services

### PR DESCRIPTION
**Pull request by Troy-Ferrell**
## Overview
This change adjust the service registration infrastructure to adhere to the original desired design. The original design involves two key concepts: 
1) The difference between the Registry & Registrar
2) The difference between IMixedRealityService & IMixedRealityDataProvider

This PR makes sure that DataProviders are not accessible by Registrars (primarily the MRTK object) and instead should be registered/accessed via the Service requiring that IMixedRealityDataProvider. In order to accomplish this:

- Deprecated all IMixedRealityDataProvider related methods in IMixedRealityServiceRegistrar
- Updated MixedRealityToolkit to not register data providers at all via RegisterService code path as well as Unregister
- Updated MixedRealityToolkit to not allow getservice on IMixedRealityDataProvider 
- Updated MixedRealityToolkit to perform service updates/coordination by requesting all services from the MixedRealityServiceRegistry
- Added BaseDataProviderAccessCoreSystem class for core systems that want to register and manage data providers.
- Updated MixedRealityInputSystem & MixedRealitySpatialAwarenessSystem to use BaseDataProviderAccessCoreSystem
- Updated MixedRealityToolkitTests to not include checks for IMixedRealityDataProvider
- Updated InputSystemTests with IMixedRealityDataProvider check
- Updated MixedRealityToolkitFacadeHandler to not use RegisteredServices property
- Discovered timing bug in ManipulationHandlerOriginOffset test
- Updated MixedRealtyServiceRegistry to handle TryGetSystems more flexibily 

## Changes
- Fixes: #5509 

## Verification
Ran all tests
Ran HandInteractionExampleScene
Ran BoundarySystem prefab in VR
Ran InputSystem prefab in Editor

> As a reviewer, it is possible to check out this change locally by using the following
> commands (substituting {PR_ID} with the ID of this pull request):
>
> git fetch origin pull/{PR_ID}/head:name_of_local_branch
>
> git checkout name_of_local_branch

### CODE REVIEW

File: Assets/MixedRealityToolkit/Services/BaseDataProviderAccessCoreSystem.cs [hunk 1]
```diff
@@ -0,0 +1,238 @@
...
+using Microsoft.MixedReality.Toolkit.Utilities;
+using System;
+
+#if UNITY_EDITOR
+using UnityEditor;
+#endif
+
+namespace Microsoft.MixedReality.Toolkit
+{
+    public abstract class BaseDataProviderAccessCoreSystem : BaseCoreSystem, IMixedRealityDataProviderAccess
```
Comment (davidkline-ms) [inline Assets/MixedRealityToolkit/Services/BaseDataProviderAccessCoreSystem.cs]: nit: can you add a summary tag for doc comments?

**Comment by davidkline-ms** [comment]
> Deprecated all IMixedRealityDataProvider related methods in IMixedRealityServiceRegistrar

These methods can and should be removed from the interface. Now is the time for the "final" breaking change.

File: Assets/MixedRealityToolkit/Interfaces/Registrars/IMixedRealityServiceRegistrar.cs [hunk 2]
```diff
@@ -85,13 +85,15 @@ public interface IMixedRealityServiceRegistrar
         /// </summary>
         /// <typeparam name="T">The interface type of the data provider to be registered.</typeparam>
         /// <param name="dataProviderInstance">An instance of the data provider to be registered.</param>
+        [Obsolete]
```
Comment (davidkline-ms) [inline Assets/MixedRealityToolkit/Interfaces/Registrars/IMixedRealityServiceRegistrar.cs]: You can remove this and related methods. Now is the time to make the last of the breaking changes and get this interface right. :)

File: Assets/MixedRealityToolkit/Services/MixedRealityToolkit.cs [hunk 3]
```diff
@@ -244,30 +244,17 @@ public void ResetConfiguration(MixedRealityToolkitConfigurationProfile profile)
...
 
-            return false;
+            return MixedRealityServiceRegistry.RemoveService<T>(serviceInstance, this);
         }
 
         /// <inheritdoc />
         public bool IsServiceRegistered<T>(string name = null) where T : IMixedRealityService
         {
-            return GetService<T>(name) != null;
+            T service;
```
Comment (davidkline-ms) [inline Assets/MixedRealityToolkit/Services/MixedRealityToolkit.cs]: nit: consider checking for IMixedRealityDataProvider and log guidance to inquire with the appropriate service.

File: Assets/MixedRealityToolkit/Services/MixedRealityToolkit.cs [hunk 4]
```diff
@@ -825,7 +797,16 @@ private bool RegisterServiceInternal(Type interfaceType, IMixedRealityService se
                 return false;
             }
 
-            if (!CanGetService(interfaceType)) { return false; }
+            if (typeof(IMixedRealityDataProvider).IsAssignableFrom(interfaceType))
+            {
+                Debug.LogWarning($"Unable to add a service of type {typeof(IMixedRealityDataProvider).Name}.");
```
Comment (davidkline-ms) [inline Assets/MixedRealityToolkit/Services/MixedRealityToolkit.cs]: nit: consider adding a note to register with the appropriate service

File: Assets/MixedRealityToolkit/Interfaces/Registrars/IMixedRealityServiceRegistrar.cs (hunk 2 above)
Comment (Troy-Ferrell) [reply Assets/MixedRealityToolkit/Interfaces/Registrars/IMixedRealityServiceRegistrar.cs]: Definitely happy to remove :)

**Comment by davidkline-ms** [comment]
@Troy-Ferrell, thanks for making these changes. They look great!

**Comment by davidkline-ms** [comment]
Test_04_RegisterMixedRealityDataProvider is failing. Likely that it is no longer appropriate with the changes that were made.

**Pull request merged**